### PR TITLE
Spyre subgraph wrapper codegen + invoke_subgraph decomp threading

### PR DESCRIPTION
**Pull request by Jiamin-Ni**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Enablement for full-graph hierarchical compile (`nested_compile_region` /
the `invoke_subgraph` HOP) on Spyre. Two independent, pure-enablement pieces —
no existing behavior changes when `nested_compile_region` is not in use.

**1. Subgraph wrapper codegen.** Splits the graph-role-agnostic parts of
`SpyrePythonWrapperCodegen` (device-layout buffer allocation, HBM-pool
reuse/free, constant-tensor fallback, sizevars patching) into
`_SpyreWrapperCodegenMixin`, and adds `SpyreSubgraphPythonWrapperCodegen`
(mixin + stock `SubgraphPythonWrapperCodegen`), now returned from
`create(is_subgraph=True)`.

Previously `create()` returned the *stock* `SubgraphPythonWrapperCodegen` for
subgraphs, so an `invoke_subgraph` body raised `AttributeError` the moment it
emitted a Spyre buffer (e.g. a `SpyreConstantFallback` materialized by
`split_multi_ops`). Role-specific behavior (header emission, benchmark harness,
launcher naming) deliberately stays on the concrete classes, so a subgraph keeps
the stock no-op `write_header` and does not re-emit imports.

**2. Decomposition threading.** `_patch_invoke_subgraph_decompositions()` wraps
`_extract_nested_region_config` so a region body that *inherits* its parent's
decompositions (config `None`) is re-traced with the Spyre decomp table.
`nested_compile_region` bodies are re-traced separately via
`reenter_make_fx(...)`, and with no table the ops that must decompose on Spyre
(notably SDPA → online-softmax) survive into the subgraph and lower incorrectly.

Gated on a new `in_spyre_compile()` thread-local (`_inductor/__init__.py`) set
for the duration of a Spyre `compile_fx`, rather than on device inspection: at
HOP re-trace time the body is traced on fake tensors whose device is not
`spyre`, so scanning the GraphModule's tensor devices always reports
"not Spyre". Thread-local (not a module global) so a concurrent CPU-only
compile on another thread cannot pick up the Spyre table. An explicit
user-supplied `NestedCompileRegionOptions(decompositions=...)` is passed through
unchanged.

#### Which issue(s) this PR is related to:

Part of the full-graph hierarchical-compile effort
https://github.com/torch-spyre/torch-spyre/issues/3738

#### Does this PR introduce a user-facing change?

No. Pure enablement — inert unless `torch.compiler.nested_compile_region` is used. It does not by itself make hierarchical compile work end-to-end; further pieces in the effort are still required.

### CODE REVIEW

**Review by dgrove-oss — APPROVED** [review]
Summary

Enables nested_compile_region/invoke_subgraph HOP compilation on Spyre via two independent pieces: (1) a SpyreSubgraphPythonWrapperCodegen wrapper class (via extracting a role-agnostic _SpyreWrapperCodegenMixin) so subgraph bodies get Spyre buffer codegen instead of crashing with AttributeError, and (2) a monkey-patch that threads the Spyre decomposition table into HOP re-traces via a new in_spyre_compile() thread-local, so ops like SDPA correctly decompose inside shared subgraph bodies. Both pieces are inert unless nested_compile_region is used.

Review

BLOCKERs

(none)

SUGGESTIONs

- [ ] Piece 2 (_patch_invoke_subgraph_decompositions) has no test coverage anywhere in this PR or the existing suite. test_subgraph_wrapper.py only covers piece 1 (wrapper codegen), and the pre-existing test_nested_compile_region_guard.py is explicitly CPU-only, so in_spyre_compile() is False throughout it and the patch is never actually exercised. Given the docstring itself describes a subtle, previously-shipped-wrong bug (SDPA lowering incorrectly once ≥2 calls share one HOP body), this is exactly the kind of fix that benefits from a regression test proving the decomp table is actually threaded through — e.g. asserting get_spyre_decomp_table() is returned from the patched _extract_nested_region_config while in_spyre_compile() is True, or a fake-tensor-level check that doesn't require a real device. Worth adding before/shortly after merge, even if full device-level end-to-end SDPA validation waits on other pieces of the hierarchical-compile effort. (torch_spyre/_monkey_patch.py:456-521)

QUESTIONs

(none)

Verdict

APPROVE

Notes: CI is green (fresh against head commit, linters pass, DCO clean on all real commits), mergeable, and the wrapper-codegen split is well-tested with clear unit tests pinning exactly the right contracts (dispatch, buffer codegen delegation, sizevars patched but header not double-emitted). The decomposition-threading patch is carefully reasoned and defensively written (idempotent, respects explicit user tables, correctly gated on a thread-local rather than device-inspection with a documented reason why device-inspection would be wrong) — but ships with no test proving it does what it claims. Since the whole PR is pure enablement inert by default, and the author is transparent that further pieces are still needed, I'm not blocking on this, but I'd want the coverage gap tracked or closed before this mechanism gets relied upon.

File: torch_spyre/_monkey_patch.py [hunk 1]
```diff
@@ -351,6 +351,81 @@ def _spyre_tensor_reuse_eval(value, metadata):
...
+    # ─────────────── invoke_subgraph subgraph decompositions ───────────────
+    # Threads the Spyre decomposition table into the re-trace of every
+    # nested_compile_region / invoke_subgraph subgraph body, so ops that must
+    # be decomposed on Spyre (notably SDPA → online-softmax) are decomposed
+    # inside the HOP body — not just in the top-level graph.
+    # ──────────────────────────────────────────────────────────────────────────
+    _patch_invoke_subgraph_decompositions()
+
+
+def _patch_invoke_subgraph_decompositions():
```
Comment (JRosenkranz) [inline torch_spyre/_monkey_patch.py:363]: Add a unit test hitting the patched ivs._extract_nested_region_config directly. Note get_spyre_decomp_table() builds a fresh dict per call, so assert by content, not identity

File: torch_spyre/_monkey_patch.py [hunk 2]
```diff
@@ -351,6 +351,81 @@ def _spyre_tensor_reuse_eval(value, metadata):
...
+    mod = sys.modules.get("torch._higher_order_ops.invoke_subgraph")
+    if mod is None:
+        try:
+            import torch._higher_order_ops.invoke_subgraph as mod  # noqa: F811
+        except ImportError:
+            # torch predates the invoke_subgraph reenter path — nothing to do.
+            return
+
+    original = getattr(mod, "_extract_nested_region_config", None)
+    if original is None or getattr(original, "_spyre_decomp_patched", False):
```
Comment (JRosenkranz) [inline torch_spyre/_monkey_patch.py:408]: Split the guard. When mod is present but original is None, that's API drift — logger.warning(...) once (the module has no logger yet, so add logger = logging.getLogger(__name__)), then return. Keep the _spyre_decomp_patched early-return separate.

**Review by JRosenkranz — APPROVED** [review]
lgtm, just minor comments that can be addressed in subsequent PRs
